crates/factor-outbound-networking/src/lib.rs

```diff
@@ -0,0 +1,116 @@
+use std::{collections::HashMap, sync::Arc};
+
+use anyhow::Context;
+use futures_util::{
+    future::{BoxFuture, Shared},
+    FutureExt,
+};
+use spin_factor_variables::VariablesFactor;
+use spin_factor_wasi::WasiFactor;
+use spin_factors::{
+    Factor, FactorInstancePreparer, InstancePreparers, PrepareContext, Result, SpinFactors,
+};
+use spin_outbound_networking::{AllowedHostsConfig, ALLOWED_HOSTS_KEY};
+
+pub struct OutboundNetworkingFactor;
+
+impl Factor for OutboundNetworkingFactor {
+    type AppConfig = AppConfig;
+    type InstancePreparer = InstancePreparer;
+    type InstanceState = ();
+
+    fn configure_app<Factors: SpinFactors>(
+        &self,
+        app: &spin_factors::App,
+        _ctx: spin_factors::ConfigureAppContext<Factors>,
+    ) -> Result<Self::AppConfig> {
+        // Extract allowed_outbound_hosts for all components
+        let component_allowed_hosts = app
+            .components()
+            .map(|component| {
+                Ok((
+                    component.id().to_string(),
+                    component
+                        .get_metadata(ALLOWED_HOSTS_KEY)?
+                        .unwrap_or_default()
+                        .into_boxed_slice()
+                        .into(),
+                ))
+            })
+            .collect::<Result<_>>()?;
+        Ok(AppConfig {
+            component_allowed_hosts,
+        })
+    }
+}
+
+#[derive(Default)]
+pub struct AppConfig {
+    component_allowed_hosts: HashMap<String, Arc<[String]>>,
+}
+
+type AllowedHostsFuture = Shared<BoxFuture<'static, Arc<anyhow::Result<AllowedHostsConfig>>>>;
+
+pub struct InstancePreparer {
+    allowed_hosts_future: AllowedHostsFuture,
+}
+
+impl FactorInstancePreparer<OutboundNetworkingFactor> for InstancePreparer {
+    fn new<Factors: SpinFactors>(
+        ctx: PrepareContext<OutboundNetworkingFactor>,
+        mut preparers: InstancePreparers<Factors>,
+    ) -> Result<Self> {
+        let hosts = ctx
+            .app_config()
+            .component_allowed_hosts
+            .get(ctx.app_component().id())
+            .cloned()
+            .context("missing component allowed hosts")?;
+        let resolver = preparers.get_mut::<VariablesFactor>()?.resolver().clone();
+        let allowed_hosts_future = async move {
+            let prepared = resolver.prepare().await?;
+            AllowedHostsConfig::parse(&hosts, &prepared)
+        }
+        .map(Arc::new)
+        .boxed()
+        .shared();
+        // let prepared_resolver = resolver.prepare().await?;
+        // let allowed_hosts = AllowedHostsConfig::parse(
+        //         .context("missing component allowed hosts")?,
+        //     &prepared_resolver,
+        // )?;
+
+        // Update Wasi socket allowed ports
+        let wasi_preparer = preparers.get_mut::<WasiFactor>()?;
```

Comment:
So are we relying on convention that app configuration is usually handled in one particular factor and we rely on that factor reaching into other factor prepares and mutating them?

In other words, why doesn't the `WasiFactor` read from `allowed_outbound_hosts` itself?

Reply:
> So are we relying on convention that app configuration is usually handled in one particular factor and we rely on that factor reaching into other factor prepares and mutating them?

In this case, yes. `allowed_outbound_hosts` supports variables, which means evaluating it requries async calls to variable provider(s) at some point. Currently spin does this eagerly at boot, which means dynamic variable changes wouldn't propagate to `allowed_outbound_hosts`. The `AllowedHostsFuture` stuff above is a first effort at resolving this issue; I expect to add some caching and/or concurrent sync in the future as well.

> In other words, why doesn't the WasiFactor read from allowed_outbound_hosts itself?

The stuff above require centralizing `allowed_outbound_hosts` resolution _somewhere_, and this is what made sense to me.